File: kotlin/vci-client/src/main/java/io/mosip/vciclient/credentialRequest/types/LdpVcCredentialRequest.kt
package io.mosip.vciclient.credentialRequest.types

import com.google.gson.annotations.SerializedName
import io.mosip.vciclient.common.JsonUtils
import io.mosip.vciclient.credentialRequest.CredentialRequest
import io.mosip.vciclient.dto.IssuerMetaData
import io.mosip.vciclient.proof.Proof
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody

class LdpVcCredentialRequest(
    override val accessToken: String,
    override val issuerMetaData: IssuerMetaData,
    override val proof: Proof,
) : CredentialRequest {
    override fun constructRequest(): Request {
        return Request.Builder()
            .url(this.issuerMetaData.credentialEndpoint)
            .addHeader("Authorization", "Bearer ${this.accessToken}")
            .addHeader("Content-Type", "application/json")
            .post(generateRequestBody())
            .build()
    }

    private fun generateRequestBody(): RequestBody {
        val credentialRequestBody = CredentialRequestBody(
            credentialDefinition = CredentialDefinition(type = this.issuerMetaData.credentialType),
            proof = proof,
            format = this.issuerMetaData.credentialFormat.value
        ).toJson()
        return credentialRequestBody
            .toRequestBody("application/json".toMediaTypeOrNull())
    }
}

private data class CredentialRequestBody(
    val format: String,
    val credentialDefinition: CredentialDefinition,
    val proof: Proof,
) {
    fun toJson(): String {
        return JsonUtils.serialize(this)
    }
}

private data class CredentialDefinition(
    @SerializedName("@context")
    val context: Array<String>? = arrayOf("https://www.w3.org/2018/credentials/v1"),
    val type: Array<String>,
)

File: js/src/credentialRequest/types/LdpVcCredentialRequest.js
const Logger = require("../../common/Logger");
const DownloadFailedException = require("../../exception/DownloadFailedException");
const CredentialRequest = require("../credentialRequest");

class LdpVcCredentialRequest extends CredentialRequest {
    constructor(accessToken, issuerMetaData, proof) {
        super();
        this.accessToken = accessToken;
        this.issuerMetaData = issuerMetaData;
        this.proof = proof;
    }

    constructRequest() {
        const request = {
          requestBody: this.generateRequestBody(),
          requestHeader: {
            'Authorization': `Bearer ${this.accessToken}`,
            'Content-Type': 'application/json',
          },
        };
        return request;
      }

    generateRequestBody() {
        try {
            const requestBody = new CredentialRequestBody(
              this.issuerMetaData.credentialFormat,
              new CredentialDefinition(this.issuerMetaData.credentialType),
              this.proof
            ).toJsonString();
            return requestBody;
        } catch (error) {
            const errorMessage = "RequestGenerationFailed";
            Logger.error(errorMessage);
            throw new DownloadFailedException(errorMessage);
        }
    }
}

class CredentialRequestBody {
    constructor(format, credentialDefinition, proof) {
        this.format = format;
        this.credential_definition = credentialDefinition;
        this.proof = proof;
    }
    toJsonString() {
        return JSON.stringify(this);
    }
}

class CredentialDefinition {
    constructor(type) {
        this.type = type;
        this["@context"] = ["https://www.w3.org/2018/credentials/v1"];
    }
}

module.exports = LdpVcCredentialRequest;